packages/core/src/oas-types.ts

```diff
@@ -1,57 +1,99 @@
+  Async2Preprocessor,
+  Async2Rule,
+} from './visitors';
+import { Oas2Types } from './types/oas2';
+import { Oas3Types } from './types/oas3';
+import { Oas3_1Types } from './types/oas3_1';
+import { AsyncApi2Types } from './types/asyncapi';
 
 export type RuleSet<T> = Record<string, T>;
 
-export enum OasVersion {
-  Version2 = 'oas2',
-  Version3_0 = 'oas3_0',
-  Version3_1 = 'oas3_1',
+export enum SpecVersion {
+  OAS2 = 'oas2',
+  OAS3_0 = 'oas3_0',
+  OAS3_1 = 'oas3_1',
+  Async2 = 'async2', // todo split into 2.x maybe?
 }
 
-export enum OasMajorVersion {
-  Version2 = 'oas2',
-  Version3 = 'oas3',
+export enum SpecMajorVersion {
+  OAS2 = 'oas2',
+  OAS3 = 'oas3',
+  Async2 = 'async2',
 }
 
+const typesMap = {
+  [SpecVersion.OAS2]: Oas2Types,
+  [SpecVersion.OAS3_0]: Oas3Types,
+  [SpecVersion.OAS3_1]: Oas3_1Types,
+  [SpecVersion.Async2]: AsyncApi2Types,
+};
+
 export type Oas3RuleSet = Record<string, Oas3Rule>;
 export type Oas2RuleSet = Record<string, Oas2Rule>;
+export type Async2RuleSet = Record<string, Async2Rule>;
 export type Oas3PreprocessorsSet = Record<string, Oas3Preprocessor>;
 export type Oas2PreprocessorsSet = Record<string, Oas2Preprocessor>;
+export type Async2PreprocessorsSet = Record<string, Async2Preprocessor>;
 export type Oas3DecoratorsSet = Record<string, Oas3Preprocessor>;
 export type Oas2DecoratorsSet = Record<string, Oas2Preprocessor>;
+export type Async2DecoratorsSet = Record<string, Async2Preprocessor>;
 
-export function detectOpenAPI(root: any): OasVersion {
+export function detectSpec(root: any): SpecVersion {
   if (typeof root !== 'object') {
     throw new Error(`Document must be JSON object, got ${typeof root}`);
   }
 
-  if (!(root.openapi || root.swagger)) {
-    throw new Error('This doesn’t look like an OpenAPI document.\n');
-  }
+  // if (!(root.openapi || root.swagger)) {
+  //   throw new Error('This doesn’t look like an OpenAPI document.\n');
+  // }
 
   if (root.openapi && typeof root.openapi !== 'string') {
     throw new Error(`Invalid OpenAPI version: should be a string but got "${typeof root.openapi}"`);
   }
 
   if (root.openapi && root.openapi.startsWith('3.0')) {
-    return OasVersion.Version3_0;
+    return SpecVersion.OAS3_0;
   }
 
   if (root.openapi && root.openapi.startsWith('3.1')) {
-    return OasVersion.Version3_1;
+    return SpecVersion.OAS3_1;
   }
 
   if (root.swagger && root.swagger === '2.0') {
-    return OasVersion.Version2;
+    return SpecVersion.OAS2;
+  }
+
+  // if not detected yet
+  if (root.openapi || root.swagger) {
+    throw new Error(`Unsupported OpenAPI Version: ${root.openapi || root.swagger}`);
+  }
+
+  if (root.asyncapi && root.asyncapi.startsWith('2.')) {
+    return SpecVersion.Async2;
+  }
+
+  if (root.asyncapi) {
+    throw new Error(`Unsupported AsyncAPI Version: ${root.asyncapi}`);
   }
 
-  throw new Error(`Unsupported OpenAPI Version: ${root.openapi || root.swagger}`);
+  throw new Error(`Unsupported specification`);
```

Comment:
Could you add a unit test that covers these cases?

Reply:
Thanks! Could you take a look at the tests I added? It looks okay locally, but should it have updated the snapshot files? It didn't .... I might need some help to get the build to pass.

Reply:
Resolved.